### PR TITLE
fix(challenges): Provide accessible test description and status

### LINKED ISSUE #13013: screen-readers don't read whether tests pass or fail
#### Browser Information

* Browser Name, Version: 
* Operating System: 
* Mobile, Desktop, or Tablet: 

#### Your Code

```js
 
 
 
```
#### Screenshot

When we're editing a challenge, we have a list of tests read above the editor, but it's impossible to check if they pass or fail. I suggest adding a textual describtion i.e. 
passed: Make sure each of your a elements has a closing tag. 
or
failed: Make sure each of your a elements has a closing tag.

### PR DESCRIPTION
**Pull request by xnt**
Closes #13013

fix(challenges): Provide accessible test description and status

#### Pre-Submission Checklist
- [X] Your pull request targets the `staging` branch of freeCodeCamp.
- [X] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [X] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [X] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change
- [X] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [X] Tested changes locally.
- [x] Closes currently open issue: Closes #13013

#### Description
* Use aria-hide on elements that aren't independently relevant for visually impaired or challenged users
* Merge test status and text on a single string
* Add the string as the aria text / label of the row element
* Force the row element to be a tab stop

Changes have been tested with Mac's VoiceOver on Chrome and Safari. Changes _should_ work on NVDA and JAWS (don't have a test environment at the moment). npm test / lint runs fine.
screen shot 2017-08-12 at 2 32 46 pm
screen shot 2017-08-12 at 2 33 24 pm

### CODE REVIEW

**Comment by QuincyLarson** [comment]
@HKuz since you're our resident accessibility expert, would you be interested in QA'ing this?

**Comment by HKuz** [comment]
I'm sorry, @QuincyLarson - I don't know anything about aria elements, I wouldn't know how to QA this.

File: common/app/routes/challenges/Test-Suite.jsx [hunk 1]
```diff
@@ -39,6 +47,24 @@ export default class TestSuite extends PureComponent {
     });
   }
 
+  getAccessibleText(err, pass, text) {
```
Comment (BerkeleyTrue) [inline common/app/routes/challenges/Test-Suite.jsx]: Why does this function need to be a method attached to the class? It should just be a top level function in this file

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

File: common/app/routes/challenges/Test-Suite.jsx (hunk 1 above)
Comment (xnt) [inline common/app/routes/challenges/Test-Suite.jsx]: @BerkeleyTrue I'm sorry. I'm not really that experienced with JSX, so I didn't think there'd be any problem with it. I've updated the pull request per your comments -making the function top-level within the file, popping it out of the class, and verifying that npm test runs nicely and the results are still the same.

Thank you!

File: common/app/routes/challenges/Test-Suite.jsx [hunk 2]
```diff
@@ -8,6 +8,24 @@ const propTypes = {
...
+function getAccessibleText(err, pass, text) {
+  let accessibleText = '';
+
+  // Determine test status (i.e. icon)
+  if (err) {
+    accessibleText += 'Error ';
+  } else if (pass) {
+    accessibleText += 'Pass';
+  } else {
+    accessibleText += 'Waiting';
```
Comment (BerkeleyTrue) [inline common/app/routes/challenges/Test-Suite.jsx]: This logic can be simplified. If Waiting is the default then this last conditional can be removed and the `accessibleText` can start out as `Waiting`.

Then the two conditionals above just overwrite the string.

File: common/app/routes/challenges/Test-Suite.jsx [hunk 3]
```diff
@@ -8,6 +8,24 @@ const propTypes = {
...
+  }
+
+  // Append the text itself
+  accessibleText += ' - ';
+  accessibleText += text;
+  return accessibleText;
```
Comment (BerkeleyTrue) [inline common/app/routes/challenges/Test-Suite.jsx]: These three lines can be condensed into one simple string concatenation.
`return accessibleText + '-' + text;`

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Comment by xnt** [comment]
@BerkeleyTrue - I've addressed your latest comments. Please take a look when you have a chance. Thank you!

**Comment by BerkeleyTrue** [comment]
@xnt It's general good practice to always test that your code changes have the desired affect. Have you done so? 

fix_challenges___provide_accessible_test_description_and_status_by_xnt_ _pull_request__15758_ _freecodecamp_freecodecamp_

This shows that the accessibility text will always have `Waiting` appended to the text. Is this the desired outcome?

Also `Error ` has a space after the word while waiting and pass do not. Is this inconsistency intended?

There is a saying in coding that goes `red light - green light`. Meaning, first validate that the issue exists (red light!), make your changes, and then validate that your code changes actually solve the issue (green light!) either by writing tests or actually running the code. I don't expect most campers to write test for new code but it should be the minimum to always run the code in the browser and verify that the outcome of your changes is what you expect.

It is also best to ask for clarification if requested changes do not make sense. Communication through the internet is always subpar and often leads to misunderstandings.

**Comment by xnt** [comment]
I find that comment rather condescending and makes me feel a bit uncomfortable, @berkeleyTrue . I'll assume that's what you mean by supbar, prone to misunderstandings. 

That's definitely a bug in my code and I'll fix it later today. That is not the intended effect.

**Comment by xnt** [comment]
Fixed, @BerkeleyTrue . Apologies for the careless mistake and not performing appropriate testing.

screen shot 2017-08-22 at 10 07 55 pm

**Comment by QuincyLarson** [comment]
@BerkeleyTrue it seems @xnt has made the requested changes (I'm not sure if he needs to do anything to the review to send you a notification through GitHub). This is ready to QA again when you get time.

**Review by BerkeleyTrue — APPROVED** [review]

**Comment by QuincyLarson** [comment]
@xnt Thanks again for your contribution!

**Comment by xnt** [comment]
@QuincyLarson my pleasure, I'll keep an eye to more a11y issues when I have a bit of downtime.

**Pull request merged**